### PR TITLE
ci: Add ignore_depends_on label

### PR DESCRIPTION
**Pull request by stevenhorsman**
- Add logic to resolve-kata-dependencies.sh to skip
depends-on processing if ignore_depends_on label
is present
- Refactor common code and tests from ci-fast-return.sh
into lib.sh

Fixes: #4360
Signed-off-by: stevenhorsman <steven@uk.ibm.com>

### CODE REVIEW

**Comment by stevenhorsman** [comment]
It's worth noting that, as we learnt on Friday evening, the code in this change gets run before the PR branch is checked out, so the testing here won't be using the new code. I'm not sure of any suggestions to minimise the risk, other than maybe merging the commit into `CCv0` first and then testing it there?

**Comment by stevenhorsman** [comment]
@jodh-intel, Sorry to bother you James, but I'd really appreciate you taking a look at this PR as I've re-used and refactored some of the code in `ci-fast-return.sh` that I think you wrote. I've tried to preserve (and add to) the unit tests, but this is the first time I'd come across `shunit2`, so there is probably a better way that I managed to get it working.

**Comment by stevenhorsman** [comment]
Manually unit testing the shell scripts:
```
$ ./ci-fast-return.sh test
INFO: Running self tests
INFO: Go get unit test framework from github.com/kward/shunit2
package github.com/kward/shunit2: no Go files in /root/go/src/github.com/kward/shunit2
INFO: Run the ci-fast-return unit tests
testYAMLreader
 0: Testing common:skip_patterns == '^CODEOWNERS$ .*\.md'
 1: Testing common:check_patterns == 'check1 .*check2$'
 2: Testing common:empty_patterns == ''
 3: Testing common:undefined_patterns == ''
 4: Testing documentation:doc_single_entry == 'single_pattern'
 5: Testing documentation:doc_complex_patterns == '^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$ ^#?([a-f0-9]{6}|[a-f0-9]{3})$'
 6: Testing nonrepo:nonrepo_pattern == ''
testPatternMatcher
testCheckForceLabel
INFO: No ghprbGhRepository set, skip label check
INFO: No forcing label found
INFO: No ghprbGhRepository set, skip label check
INFO: No forcing label found
INFO: No ghprbGhRepository set, skip label check
INFO: No forcing label found
INFO: BUG: No label to check specified
INFO: No forcing label found
INFO: label 'force-ci' not found
INFO: No forcing label found
INFO: label 'force-ci' found
INFO: Forcing CI
testCheckSkipLabel
INFO: No ghprbGhRepository set, skip label check
INFO: No CI skip label found
INFO: No ghprbGhRepository set, skip label check
INFO: No CI skip label found
INFO: No ghprbGhRepository set, skip label check
INFO: No CI skip label found
INFO: BUG: No label to check specified
INFO: No CI skip label found
INFO: label 'force-skip-ci' not found
INFO: No CI skip label found
INFO: label 'force-skip-ci' found
INFO: Skipping CI
testIsLabelSet
INFO: BUG: No repo specified
INFO: BUG: No PR specified
INFO: BUG: No label to check specified
INFO: BUG: No repo specified
INFO: BUG: No PR specified
INFO: Checking labels for PR repo/pr
testCheckLabel
INFO: BUG: No label to check specified
INFO: No ghprbGhRepository set, skip label check
INFO: No ghprbGhRepository set, skip label check
INFO: No ghprbGhRepository set, skip label check
INFO: label 'label' not found
INFO: label 'label' found

Ran 6 tests.

OK
```
```
$ ./resolve-kata-dependencies.sh test
INFO: Running self tests
INFO: Go get unit test framework from github.com/kward/shunit2
package github.com/kward/shunit2: no Go files in /root/go/src/github.com/kward/shunit2
INFO: Run the unit tests
testCheckIgnoreDependsOnLabel
INFO: No ghprbGhRepository set, skip label check
INFO: No ignore_depends_on label found
INFO: No ghprbGhRepository set, skip label check
INFO: No ignore_depends_on label found
INFO: No ghprbGhRepository set, skip label check
INFO: No ignore_depends_on label found
INFO: BUG: No label to check specified
INFO: No ignore_depends_on label found
INFO: label 'ignore-depends-on' not found
INFO: No ignore_depends_on label found
INFO: label 'ignore-depends-on' found
INFO: Ignoring all 'Depends-on' labels
testIsLabelSet
INFO: BUG: No repo specified
INFO: BUG: No PR specified
INFO: BUG: No label to check specified
INFO: BUG: No repo specified
INFO: BUG: No PR specified
INFO: Checking labels for PR repo/pr
testCheckLabel
INFO: BUG: No label to check specified
INFO: No ghprbGhRepository set, skip label check
INFO: No ghprbGhRepository set, skip label check
INFO: No ghprbGhRepository set, skip label check
INFO: label 'label' not found
INFO: label 'label' found

Ran 3 tests.

OK
```

**Review by jodh-intel — DISMISSED** [review]
Thanks @stevenhorsman !

*Quis custodiet ipsos custodes?* Love it! :smile:

**Comment by stevenhorsman** [comment]
My guess was correct. It's quite confusing (to me at least), but the `kata_repo` variable is used generally for either the tests repo, or kata-containers repo, depending on which one we are testing, then `katacontainers_repo="${katacontainers_repo:-github.com/kata-containers/kata-containers}"` and 
`tests_repo="${tests_repo:-github.com/kata-containers/tests}"` are used for the specific two repos. So for this tests PR `kata_repo=github.com/kata-containers/tests`.

However in `lib.sh` it sets `export kata_repo=${katacontainers_repo:="github.com/kata-containers/kata-containers"}`, so after that is called the `kata_repo` is overwritten with the wrong repo. I'm going to attempt to remove this inconsistency, so see if that helps, and if I can get everything else aligned. 🤞

**Comment by stevenhorsman** [comment]
I've added a commit with my first attempt at making the repo names consistent and am now starting the testing...

File: .ci/ci-fast-return.sh
Comment (Jakob-Naucke) [inline .ci/ci-fast-return.sh:421]: Can't we just `shunit2` `lib.sh`?

File: .ci/resolve-kata-dependencies.sh
Comment (Jakob-Naucke) [inline .ci/resolve-kata-dependencies.sh:23]: I know we're not exactly minimal in a bunch of places, but are we using `os-release` for anything here?

File: .ci/ci-fast-return.sh
Comment (stevenhorsman) [reply .ci/ci-fast-return.sh:421]: Hey @Jakob-Naucke - I might have misunderstood you, but I originally tried adding `"${GOPATH}/src/${shunit2_path}/shunit2" lib.sh` to the `self_test`

and I got:
```
shunit2:FATAL unable to read from lib.sh
```
I'm guessing I did something wrong there?

File: .ci/resolve-kata-dependencies.sh
Comment (stevenhorsman) [reply .ci/resolve-kata-dependencies.sh]: Agreed, I don't use it, I probably get copied too many lines over. I've removed this line now. Thanks!

File: .ci/ci-fast-return.sh
Comment (Jakob-Naucke) [reply .ci/ci-fast-return.sh:421]: I meant: Can't we just make `lib.sh` another file that has self-tests, and not test these functions from `ci-fast-return.sh`/`resolve-kata-dependencies.sh`?

Comment (stevenhorsman) [reply .ci/ci-fast-return.sh:421]: Sorry for misunderstanding. 

The reason I stayed away from that approach was that I wasn't sure if we wanted a `main` function (which would execute the shunit tests if the `test` argument was applied) added to `lib.sh` as it is a common library script. It has a bunch of code that runs outside of its functions and is sourced in a lot of places, so we'd end up executing main a lot of times (which would do nothing if it didn't have parameters).

I'm happy to give it a go though if you think it is worth trying.

Comment (Jakob-Naucke) [reply .ci/ci-fast-return.sh:421]: No need to apologise, I was too brief :)
DYT something like [`[ "$0" != "$BASH_SOURCE" ]`](https://stackoverflow.com/a/23009039) would work for introducing a main function? This slightly resembles Python's `if __name__ == "__main__"`, which is only true when the file itself is executed, but not when imported.

Comment (stevenhorsman) [reply .ci/ci-fast-return.sh]: > DYT something like [`[ "$0" != "$BASH_SOURCE" ]`](https://stackoverflow.com/a/23009039) would work for introducing a main function? This slightly resembles Python's `if __name__ == "__main__"`, which is only true when the file itself is executed, but not when imported.

I really like that idea. I've had an attempt at it in https://github.com/kata-containers/tests/pull/4363/commits/a48e644b740ff862146279f6aa6bb1cc846b1c10. If you are happy with it and all the test look okay then I'll squash it into the unit tests commit, but I've kept it separate for now whilst this bit is 'in progress'

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

File: .ci/lib.sh
Comment (Jakob-Naucke) [inline .ci/lib.sh:755]: I think this would be inverse, this is true _when_ `source`d. (Did this work?!)

Comment (stevenhorsman) [reply .ci/lib.sh:755]: I'm confused. Here is a little test script I made:
```bash
echo "\$0: $0"
echo "\$BASH_SOURCE: $BASH_SOURCE"
[ "$0" == "$BASH_SOURCE" ] && echo "not sourced"
```
and the results of me running it not sourced:
```
$ ./test.sh
$0: ./test.sh
$BASH_SOURCE: ./test.sh
not sourced
```
and sourced:
```
$ . test.sh
$0: -bash
$BASH_SOURCE: test.sh
```

Comment (Jakob-Naucke) [reply .ci/lib.sh:755]: Ah no, okay, had a misunderstanding about how `$BASH_SOURCE` works. Thanks for checking!

**Review by Jakob-Naucke — APPROVED** [review]
LGTM, thanks @stevenhorsman! IMO you can squash (which invalidates this approval but whatever)

**Review by Jakob-Naucke — DISMISSED** [review]

**Comment by stevenhorsman** [comment]
@jodh-intel - are you able to take another look at this PR after I've added the updated from Jakob. TYVM!

**Comment by stevenhorsman** [comment]
I've just spotted in the test output:
```
/tmp/jenkins/workspace/kata-containers-2.0-tests-ubuntu20.04-PR/go/src/github.com/kata-containers/tests/.ci/lib.sh: line 606: jq: command not found
INFO: label 'ignore-depends-on' not found
```
so I need to debug why install_jq isn't working the first time it runs

**Comment by stevenhorsman** [comment]
It looks to be working now. With the label on this PR, the [ubuntu job](http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu20.04-PR/1067/consoleFull) shows:
```
15:09:17 INFO: label 'ignore-depends-on' found
15:09:17 INFO: Ignoring all 'Depends-on' labels
15:09:17 INFO: Not applying depends on due to 'ignore-depends-on' label
```
when the rest of the jobs have finished I'll try running it without the label set to check that case too...

**Comment by stevenhorsman** [comment]
Without the label were correctly get `INFO: No ignore_depends_on label found` logged in the job output.
All the platforms are passing except centos (its daily baseline is also broken) and s390x, which is failing the stability tests with:
```
Wrong number of qemu running (0 != 1) - stopping
```
I don't think that failure is caused by this PR though.

**Comment by stevenhorsman** [comment]
@Jakob-Naucke & @jodh-intel - are you able to re-review this PR please after your original one went stale? Thank you!

**Review by sameo — APPROVED** [review]

**Review by Jakob-Naucke — APPROVED** [review]

**Review by jodh-intel — APPROVED** [review]
Thanks @stevenhorsman.

lgtm

**Pull request merged**
